From de2e00a99ba2e32132d634c000f4c833fd0aebcd Mon Sep 17 00:00:00 2001
From: Christian Hohnstaedt <christian@hohnstaedt.de>
Date: Fri, 28 May 2021 23:42:29 +0200
Subject: [PATCH] s6-linux-init: add as new recipe

---
 src/init/s6-linux-init.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/src/init/s6-linux-init.c b/src/init/s6-linux-init.c
index bb24a52..7538bb2 100644
--- a/src/init/s6-linux-init.c
+++ b/src/init/s6-linux-init.c
@@ -35,6 +35,7 @@
 #define BANNER "\n  s6-linux-init version " S6_LINUX_INIT_VERSION "\n\n"
 
 static int inns = 0 ;
+static int nokbd = 1 ;
 static int nologger = 0 ;
 static int notifpipe[2] ;
 
@@ -70,7 +71,7 @@ static inline void wait_for_notif (int fd)
 static void kbspecials (void)
 {
   int fd ;
-  if (inns) return ;
+  if (inns || nokbd) return ;
   fd = open("/dev/tty0", O_RDONLY | O_NOCTTY) ;
   if (fd < 0)
     strerr_warnwu2sys("open /dev/", "tty0 (kbrequest will not be handled)") ;
@@ -140,7 +141,7 @@ int main (int argc, char const **argv, char const *const *envp)
     subgetopt l = SUBGETOPT_ZERO ;
     for (;;)
     {
-      int opt = subgetopt_r(argc, argv, "c:p:s:m:d:D:nNCB", &l) ;
+      int opt = subgetopt_r(argc, argv, "c:p:s:m:d:D:nNCBK", &l) ;
       if (opt == -1) break ;
       switch (opt)
       {
@@ -154,6 +155,7 @@ int main (int argc, char const **argv, char const *const *envp)
         case 'N' : mounttype = 0 ; break ;
         case 'C' : inns = 1 ; break ;
         case 'B' : nologger = 1 ; break ;
+        case 'K' : nokbd = 1 ; break ;
         default : dieusage() ;
       }
     }
