From 34f0180dedb476d7f4f9a0822e871f81f26945c0 Mon Sep 17 00:00:00 2001
From: Christian Hohnstaedt <christian@hohnstaedt.de>
Date: Thu, 29 Sep 2022 10:49:58 +0200
Subject: [PATCH] s6-log: add P option as message prefix after the timestamp

---
 src/daemontools-extras/s6-log.c | 49 +++++++++++++++++++--------------
 1 file changed, 28 insertions(+), 21 deletions(-)

diff --git a/src/daemontools-extras/s6-log.c b/src/daemontools-extras/s6-log.c
index 5a7ce4d..bc2c277 100644
--- a/src/daemontools-extras/s6-log.c
+++ b/src/daemontools-extras/s6-log.c
@@ -128,6 +128,7 @@ struct act_s
   acttype_t type ;
   actstuff_t data ;
   unsigned int flags ;
+  stralloc prefix ;
 } ;
 
 typedef struct scriptelem_s scriptelem_t, *scriptelem_t_ref ;
@@ -713,6 +714,7 @@ static inline void script_firstpass (char const *const *argv, unsigned int *sell
       case 'l' :
       case 'r' :
       case 'E' :
+      case 'P' :
       case '^' :
 #ifdef S6_USE_EXECLINE
       case '!' :
@@ -766,6 +768,7 @@ static inline void script_firstpass (char const *const *argv, unsigned int *sell
 static inline void script_secondpass (char const *const *argv, scriptelem_t *script, sel_t *selections, act_t *actions)
 {
   tain retrytto ;
+  stralloc prefix = STRALLOC_ZERO ;
   unsigned int fd2_size = 200 ;
   unsigned int status_size = 1001 ;
   uint32_t s = 99999 ;
@@ -855,31 +858,35 @@ static inline void script_secondpass (char const *const *argv, scriptelem_t *scr
       case 'T' :
         flags |= 2 ;
         break ;
+      case 'P' :
+        if (!stralloc_copys(&prefix, *argv +1)) dienomem() ;
+        if (!stralloc_append(&prefix, ' ')) dienomem() ;
+        break ;
       case '1' :
       {
-        act_t a = { .type = ACTTYPE_FD1, .flags = flags } ;
-        actions[act++] = a ; flagacted = 1 ; flags = 0 ;
+        act_t a = { .type = ACTTYPE_FD1, .flags = flags, .prefix = prefix };
+        actions[act++] = a ; flagacted = 1 ; flags = 0 ; prefix = stralloc_zero ;
         break ;
       }
       case '2' :
       {
-        act_t a = { .type = ACTTYPE_FD2, .flags = flags, .data = { .fd2_size = fd2_size } } ;
-        actions[act++] = a ; flagacted = 1 ; flags = 0 ;
+        act_t a = { .type = ACTTYPE_FD2, .flags = flags, .data = { .fd2_size = fd2_size }, .prefix = prefix } ;
+        actions[act++] = a ; flagacted = 1 ; flags = 0 ; prefix = stralloc_zero ;
         break ;
       }
       case '=' :
       {
-        act_t a = { .type = ACTTYPE_STATUS, .flags = flags, .data = { .status = { .file = *argv + 1, .filelen = status_size } } } ;
-        actions[act++] = a ; flagacted = 1 ; flags = 0 ;
+        act_t a = { .type = ACTTYPE_STATUS, .flags = flags, .data = { .status = { .file = *argv + 1, .filelen = status_size } }, .prefix = prefix } ;
+        actions[act++] = a ; flagacted = 1 ; flags = 0 ; prefix = stralloc_zero ;
         break ;
       }
       case '.' : 
       case '/' :
       {
-        act_t a = { .type = ACTTYPE_DIR, .flags = flags, .data = { .ld = lidx } } ;
+        act_t a = { .type = ACTTYPE_DIR, .flags = flags, .data = { .ld = lidx}, .prefix = prefix } ;
         logdir_init(lidx, s, n, tolerance, maxdirsize, &retrytto, processor, *argv, flags) ;
         lidx++ ;
-        actions[act++] = a ; flagacted = 1 ; flags = 0 ;
+        actions[act++] = a ; flagacted = 1 ; flags = 0 ; prefix = stralloc_zero ;
         break ;
       }
       default : goto fail ;
@@ -949,11 +956,11 @@ static void script_run (scriptelem_t const *script, unsigned int scriptlen, char
       for (j = 0 ; j < script[i].actlen ; j++)
       {
         act_t const *act = script[i].acts + j ;
-        struct iovec v[4] = { { .iov_base = tstamp, .iov_len = act->flags & 1 ? TIMESTAMP+1 : 0 }, { .iov_base = hstamp, .iov_len = act->flags & 2 ? hlen : 0 }, { .iov_base = (char *)s, .iov_len = len }, { .iov_base = "\n", .iov_len = 1 } } ;
+        struct iovec v[5] = { { .iov_base = tstamp, .iov_len = act->flags & 1 ? TIMESTAMP+1 : 0 }, { .iov_base = hstamp, .iov_len = act->flags & 2 ? hlen : 0 }, { .iov_base = act->prefix.s, .iov_len = act->prefix.len }, { .iov_base = (char *)s, .iov_len = len }, { .iov_base = "\n", .iov_len = 1 } } ;
         switch (act->type)
         {
           case ACTTYPE_FD1 :
-            if (!bufalloc_putv(bufalloc_1, v, 4)) dienomem() ;
+            if (!bufalloc_putv(bufalloc_1, v, 5)) dienomem() ;
           case ACTTYPE_NOTHING :
             break ;
 
@@ -962,38 +969,38 @@ static void script_run (scriptelem_t const *script, unsigned int scriptlen, char
             buffer_puts(buffer_2, ": alert: ") ;
             if (act->data.fd2_size && act->data.fd2_size + 3 < len)
             {
-              v[2].iov_len = act->data.fd2_size ;
-              v[3].iov_base = "...\n" ;
-              v[3].iov_len = 4 ;
+              v[3].iov_len = act->data.fd2_size ;
+              v[4].iov_base = "...\n" ;
+              v[4].iov_len = 4 ;
             }
-            buffer_putv(buffer_2, v, 4) ;
+            buffer_putv(buffer_2, v, 5) ;
             buffer_flush(buffer_2) ; /* if it blocks, too bad */
             break ;
 
           case ACTTYPE_STATUS :
             if (act->data.status.filelen)
             {
-              size_t reallen = siovec_len(v, 4) ;
+              size_t reallen = siovec_len(v, 5) ;
               if (reallen > act->data.status.filelen)
-                siovec_trunc(v, 4, act->data.status.filelen) ;
+                siovec_trunc(v, 5, act->data.status.filelen) ;
               else
               {
                 size_t k = act->data.status.filelen - reallen + 1 ;
                 char pad[k] ;
-                v[3].iov_base = pad ;
-                v[3].iov_len = k ;
+                v[4].iov_base = pad ;
+                v[4].iov_len = k ;
                 while (k--) pad[k] = '\n' ;
-                if (!openwritevnclose_suffix(act->data.status.file, v, 4, ".new") && verbosity)
+                if (!openwritevnclose_suffix(act->data.status.file, v, 5, ".new") && verbosity)
                   strerr_warnwu2sys("write status file ", act->data.status.file) ;
                 break ;
               }
             }
-            if (!openwritevnclose_suffix(act->data.status.file, v, 4, ".new") && verbosity)
+            if (!openwritevnclose_suffix(act->data.status.file, v, 5, ".new") && verbosity)
               strerr_warnwu2sys("write status file ", act->data.status.file) ;
             break ;
 
           case ACTTYPE_DIR :
-            if (!bufalloc_putv(&logdirs[act->data.ld].out, v, 4)) dienomem() ;
+            if (!bufalloc_putv(&logdirs[act->data.ld].out, v, 5)) dienomem() ;
             break ;
 
           default :
